chore: add api-docs package

sissbruecker opened the pull request:
## Description

Adds an Astro-based API docs website that is intended to replace the current `iron-docs-viewer` based one at some point.

Compared to the current one, it has the following benefits:
- It's based on Astro and generates static HTML. That should make it more useful for search engines and AI tools. In comparison, the current API docs are basically an empty HTML page with a single custom element on it and AI assistants (at least GH Copilot) can't find any info on it when you pass them a URL.
- It's still based on the Polymer analysis JSON, but instead of using the type annotations from the JSDoc it looks up types from the Typescript declaration files. It also inlines all custom type definitions used in a component (events, renderers, i18n objects) into the page for easier reference.
- It generates static Markdown files as well, which are easier to process by AI tools. Each component page has a link to that MD file, as well as a copy button that allows to just copy+paste that MD as context into a prompt. Adding an index of these pages to comply with the llms.txt schema would also be possible.
- Apart from that it has the usual affordances from Astro like a search function or theme switcher.

## Type of change

- Feature

sissbruecker commented on api-docs/generate.ts:
```diff
@@ -0,0 +1,186 @@
+function sanitizeDescription(description: string) {
+  // Fix links to other elements
+  description = description.replace(/#\/elements\//gu, '/elements/');
+  // Fix links to properties (e.g. #property-dataProvider to #dataprovider)
```
Document fragments might need some work, I think there are some duplicates on some pages at the moment. Could be improved later IMO, would require figuring out how to customize these in Astro from within the Markdown file.

vursen replied:
You mean such duplicates? Or is it a different issue?

<img width="1512" height="1600" alt="image" src="https://github.com/user-attachments/assets/ae940820-fd38-444a-a51d-a819fbda58ad" />

web-padawan replied:
For the record, the new 11ty based version handles this by changing the second heading to `#dataprovider-1`.

sissbruecker commented on api-docs/generate.ts:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  // Properties
+  const publicProperties = filterPublicApi(element.properties);
```
The new API docs only show public APIs.

sissbruecker commented on api-docs/generate.ts:
```diff
@@ -0,0 +1,186 @@
+    const mdContent = renderElement(element);
+
+    try {
+      writeFileSync(docFile, mdContent, 'utf-8');
+      writeFileSync(plainTextFile, mdContent, 'utf-8');
```
The first file is a "content" file with `.md` extension that Astro will render as HTML.

The second one is a static file with `.txt` extension that contains the plain Markdown. Astro will just treat these as static assets and return them as plain text.

sissbruecker replied:
I was concerned about using `.md` because the content files have the same extension. But it seems using the same extension works fine, so `/elements/vaadin-button` renders the content one, while `/elements/vaadin-button.md` returns the one from the static folder. This will also help with implementing the [llms.txt](https://llmstxt.org/) which uses the `.md` extension in the URL as well.

sissbruecker commented on api-docs/package.json:
```diff
@@ -0,0 +1,23 @@
+    "build": "astro build",
+    "preview": "astro preview",
+    "astro": "astro",
+    "generate": "tsx generate.ts"
```
Used `tsx` for now. Since Node 23 this works without, but we don't have that on CI servers yet.

sissbruecker commented on api-docs/src/components/CustomPageTitle.astro:
This customizes the default page header to have buttons / links for the Markdown content.

sissbruecker commented on api-docs/src/content.config.ts:
```diff
@@ -0,0 +1,14 @@
+    loader: docsLoader(),
+    schema: docsSchema({
+      extend: z.object({
+        pageType: z.optional(z.string()),
```
Customizes the page schema to add `pageType`. This is used to only render the Markdown button / links on element pages, but not on the index page.

sissbruecker commented on api-docs/types.ts:
This implements the type lookup from the TS declaration files.

sissbruecker commented on api-docs/generate.ts:
This generates the content for the Astro site based on `analysis.json`.

sissbruecker commented on tsconfig.build.json:
```diff
@@ -6,7 +6,8 @@
     "lib": ["esnext", "es2018", "dom"],
     "declaration": true,
     "strict": true,
-    "skipDefaultLibCheck": true
```
Astro brings in `@types/mdx`, which depends on the `JSX` namespace being defined. We can enable `skipLibCheck` here, but not in `tsconfig.json` apparently. Needs some solution to make the JSX namespace available.

web-padawan replied:
This file is now removed in https://github.com/vaadin/web-components/pull/9921

sissbruecker replied:
Removed this change

vursen commented:
On some pages, there are duplicate event names that differ only by a hyphen:

![Screenshot 2025-08-08 at 12 04 32](https://github.com/user-attachments/assets/6d6717be-392d-47c2-8e8f-4f92504d750c)

sissbruecker commented:
> On some pages, there are duplicate event names that differ only by a hyphen

For some reason I added generation of property change events, I could have sworn some were missing from the JSON. But after double-checking that doesn't seem to be the case. So I removed that part of the generator. Looks like the number of events now line up between old and new API docs.

sissbruecker commented on api-docs/package.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "api-docs",
+  "version": "25.0.0-alpha8",
```
Thanks, done
